racket-cas/racket-cas.rkt

```diff
@@ -848,174 +880,185 @@
-              [_ #f]))
-
-(define memoized-term-with-negative-coeff? (memoize term-with-negative-coeff?))
-(module+ test
-  (check-equal? (memoized-term-with-negative-coeff? '(* -2 x log 3)) #t)
-  (check-equal? (memoized-term-with-negative-coeff? 1+i) #f)
-  (check-equal? (memoized-term-with-negative-coeff? '(+ x -10)) #f)
-  )
 
-; (positive negative)
-(define (partition-positive/negative us)
-  (when debugging? (displayln (list 'partition-positive/negative us)))
-  (define-values (n d) (partition (negate memoized-term-with-negative-coeff?) us))
-  (define (reduce-lst us vs)
-    (values (apply ⊕ us) (⊖ (apply ⊕ vs)))
-    )
-  (reduce-lst n d)
-  )
+;;;
+;;; TERM MANIPULATION
+;;;
 
-; Partition a Sum into positive and negative
-(define (positive/negative s)
-  (when debugging? (displayln (list 'positive/negative s)))
-  (math-match s
-              [(⊖ u) (values 0 u)]
-              [(Sum us) (partition-positive/negative us)]
-              [_ (values s 0)]))
+;;; Negative Term
+
+; A term of the form
+;   r or (* r u)
+; where r is a negative number is wille be called a *negative-term*.
+
+(define (negative-term? u)
+  ; we assume u is normalized, so we don't need to look at other
+  ; factors besides the first one.
+  (math-match u
+    [r       #:when (negative? r) #t]
+    [(⊗ r v) #:when (negative? r) #t]
+    [_                            #f]))
 
 (module+ test
-  (let-values ([(p n)
-               (positive/negative (normalize '(+ (Exp 5) (* -2 x) z 2/3)))])
-    (check-equal? p '(+ 2/3 z (Exp 5)))
-    (check-equal? n '(* 2 x))
-    )
-  )
+  (displayln "TEST - negative-term?")
+  (check-equal? (negative-term? (normalize '(* -2 x (log 3)))) #t)
+  (check-equal? (negative-term? (normalize '(* x (* -2 y) z))) #t)
+  (check-equal? (negative-term? (normalize 1+i))               #f)
+  (check-equal? (negative-term? (normalize '(+ x -10)))        #f))
+
+
+;;; Partition a sum into positive and negative terms.
+
+(define (partition-terms-after-sign u)
+  ; Given a sum u, two normalized expressions u_pos and u_neg such
+  ; that u = u_pos - u_neg are returned.
+  ; The sum u_neg consists of all negative terms in u (with signs negated).
+
+  ; Implementation note: If u is normalized, then so will a partial sum,
+  ; if we keep the original order.  
+  (define (reverse-terms->sum vs)    
+    (match vs
+      ['()      0]
+      [(list v) v]
+      [_        (cons '+ (reverse vs))])) ; a partial sum is normalized
+  (define (negate-term v)
+    (math-match v
+      [r          (- r)]
+      [(⊗ r u) (⊗ (- r) u)]
+      [_ (error)]))
+  (match u
+    [r #:when (and (real? r) (negative? r)) (values 0 (- r))]
+    [r #:when      (real? r)                (values r 0)]
+    [x #:when (symbol? x)                   (values x 0)] ; assume variables are positive
+    [(list* '+ us) (let loop ([us us] [pos '()] [neg '()])
+                     (match us
+                       ['()         (values (reverse-terms->sum pos)
```

Comment:
Why reverse pos? It is already in normalized order?

Reply:
The loop accumulates the terms in pos and neg in reverse order.